README.md
# RADIUS-PKIX-CD
RADIUS plugin. Uses DANE PKIX-CD mode for enabling EAP-TLS


This is incomplete... currently sorting out integration testing of EAP-TLS client against RADIUS

Most likely an issue with OpenSSL locating signing certs for handshake.

CHANGELOG.rst
Changelog
=========


v0.11
-----
- Build(deps): bump dane-discovery from 0.19 to 0.20. [dependabot[bot]]

  Bumps [dane-discovery](https://github.com/valimail/dane_discovery) from 0.19 to 0.20.
  - [Release notes](https://github.com/valimail/dane_discovery/releases)
  - [Commits](https://github.com/valimail/dane_discovery/commits)

  ---
  updated-dependencies:
  - dependency-name: dane-discovery
    dependency-type: direct:production
    update-type: version-update:semver-minor
  ...


v0.10 (2021-10-23)
------------------

Fix
~~~
- Correct structure for json file format. [Ash Wilson]


v0.9 (2021-10-22)
-----------------

Changes
~~~~~~~
- Authorize certificates by hash of entity certificate, instead of AKI.
  [Ash Wilson]

  Close #7


v0.8 (2021-10-21)
-----------------

New
~~~
- Add support for iotregistry.ca. [Ash Wilson]


v0.7 (2021-10-18)
-----------------
- Build(deps): bump dane-discovery from 0.18 to 0.19. [dependabot[bot]]

  Bumps [dane-discovery](https://github.com/valimail/dane_discovery) from 0.18 to 0.19.
  - [Release notes](https://github.com/valimail/dane_discovery/releases)
  - [Commits](https://github.com/valimail/dane_discovery/commits)

  ---
  updated-dependencies:
  - dependency-name: dane-discovery
    dependency-type: direct:production
    update-type: version-update:semver-minor
  ...


v0.6 (2021-10-02)
-----------------

Changes
~~~~~~~
- Update testing configuration and OS dependencies. [Ash Wilson]

Other
~~~~~
- Build(deps): bump dane-discovery from 0.17 to 0.18. [dependabot[bot]]

  Bumps [dane-discovery](https://github.com/valimail/dane_discovery) from 0.17 to 0.18.
  - [Release notes](https://github.com/valimail/dane_discovery/releases)
  - [Commits](https://github.com/valimail/dane_discovery/commits)

  ---
  updated-dependencies:
  - dependency-name: dane-discovery
    dependency-type: direct:production
    update-type: version-update:semver-minor
  ...


v0.5 (2021-07-21)
-----------------

New
~~~
- Support PKI with intermediate CA certificates. [Ash Wilson]

  Close #1

Changes
~~~~~~~
- Increment minor version, update CHANGELOG.rst. [Ash Wilson]

Other
~~~~~
- Build(deps): bump dane-discovery from 0.16 to 0.17. [dependabot[bot]]

  Bumps [dane-discovery](https://github.com/valimail/dane_discovery) from 0.16 to 0.17.
  - [Release notes](https://github.com/valimail/dane_discovery/releases)
  - [Commits](https://github.com/valimail/dane_discovery/commits)

  ---
  updated-dependencies:
  - dependency-name: dane-discovery
    dependency-type: direct:production
    update-type: version-update:semver-minor
  ...


v0.2 (2021-06-05)
-----------------
- Initial commit. [Ash]




tests/scripts/generate_crypto.sh
#!/bin/bash
set -xe
###################################
# This phase creates all the certs
# and keys we'll use in a later
# phase.
###################################

##############
# Set env vars
##############
# Domains
LOCAL_DOMAIN="example.net"
IOTREG_DOMAIN="iotregistry.ca"

# Base paths
CRYPTO_EXPORT_PATH="${HOME}/export"
CRYPTO_DIR="${HOME}/crypto"
LOCAL_CA_DIR="${CRYPTO_DIR}/local_ca"
DEV_DIR="${CRYPTO_DIR}/device"

# General device info
RSA_DEVICE_SERIAL="rsa"
DEVICE_MODEL="air-quality-sensor"
ECC_DEVICE_SERIAL="ecc"
IOTREG_DEVICE_SERIAL="iotreg"

# Building DIDN-IDs for non-registry devices
RSA_DIDN_ID="${RSA_DEVICE_SERIAL}.${DEVICE_MODEL}._device.${LOCAL_DOMAIN}"
ECC_DIDN_ID="${ECC_DEVICE_SERIAL}.${DEVICE_MODEL}._device.${LOCAL_DOMAIN}"

# Building CN and DIDN_ID for registry-bound devices
IOTREG_CN="${IOTREG_DEVICE_SERIAL}.${IOTREG_DOMAIN}"
IOTREG_DIDN_ID="${IOTREG_DEVICE_SERIAL}._device.${LOCAL_DOMAIN}"

# Specific file paths

## Local CA
LOCAL_CA_KEY="${LOCAL_CA_DIR}/ca.example.net.key.pem"
LOCAL_CA_CERT="${LOCAL_CA_DIR}/ca.example.net.cert.pem"

## RSA Dev
RSA_DEV_KEY="${DEV_DIR}/${RSA_DIDN_ID}.key.pem"
RSA_DEV_CSR="${DEV_DIR}/${RSA_DIDN_ID}.csr.pem"
RSA_DEV_CERT="${DEV_DIR}/${RSA_DIDN_ID}.cert.pem"

## ECC Dev
ECC_DEV_KEY="${DEV_DIR}/${ECC_DIDN_ID}.key.pem"
ECC_DEV_CSR="${DEV_DIR}/${ECC_DIDN_ID}.csr.pem"
ECC_DEV_CERT="${DEV_DIR}/${ECC_DIDN_ID}.cert.pem"

## IOTREG Dev
IOTREG_DEV_KEY="${DEV_DIR}/iotreg.ca.key.pem"
IOTREG_DEV_CSR="${DEV_DIR}/iotreg.ca.csr.pem"
IOTREG_DEV_CERT="${DEV_DIR}/iotreg.ca.cert.pem"

# Signing configurations
RSA_SSL_CONFIG="${DEV_DIR}/${RSA_DIDN_ID}.conf"
ECC_SSL_CONFIG="${DEV_DIR}/${ECC_DIDN_ID}.conf"

mkdir -p ${LOCAL_CA_DIR}/demoCA/
echo "05" > ${LOCAL_CA_DIR}/demoCA/serial 


sudo cp tests/configs/ssl/ssl.cnf /usr/lib/ssl/openssl.cnf
# cat /usr/lib/ssl/openssl.cnf



##############
# Create dirs
##############
mkdir -p \
    ${CRYPTO_EXPORT_PATH} \
    ${CRYPTO_DIR} \
    ${LOCAL_CA_DIR} \
    ${DEV_DIR} \
    ${LOCAL_CA_DIR}/demoCA/

##############
# Create local
# CA
##############
cd ${LOCAL_CA_DIR}
openssl genrsa \
    -out ${LOCAL_CA_KEY} \
    4096
openssl req \
    -key ${LOCAL_CA_KEY} \
    -new \
    -x509 \
    -days 7300 \
    -sha256 \
    -subj "/C=US/ST=CA/O=Example Networks/CN=${CA_CN:-NADA}" \
    -out ${LOCAL_CA_CERT}
openssl x509 -noout -text -in ${LOCAL_CA_CERT}
touch ${LOCAL_CA_DIR}/demoCA/index.txt
touch ${LOCAL_CA_DIR}/demoCA/index.txt.attr

##############
# Create local
# device RSA 2048
##############
cd ${LOCAL_CA_DIR}
openssl genrsa \
    -out ${RSA_DEV_KEY} \
    2048
openssl req \
    -key ${RSA_DEV_KEY} \
    -new \
    -sha256 \
    -subj "/C=US/ST=CA/O=Example Networks/CN=${RSA_DIDN_ID}" \
    -addext "subjectAltName = DNS:${RSA_DIDN_ID}" \
    -addext "keyUsage = nonRepudiation, digitalSignature, keyEncipherment" \
    -out ${RSA_DEV_CSR}

echo "#################### LOCAL RSA DEV CSR ####################"
openssl req -noout -text -in ${RSA_DEV_CSR}
# Add the SAN config to the end of the openssl conf file.
cp /usr/lib/ssl/openssl.cnf ${RSA_SSL_CONFIG}
echo -e "\n[alternate_names]" >> ${RSA_SSL_CONFIG}
echo -e "DNS.1 = ${RSA_DIDN_ID}\n" >> ${RSA_SSL_CONFIG}
# Accommodate the default behavior of openssl ca.
openssl ca \
  -config ${RSA_SSL_CONFIG} \
  -days 375 \
  -in ${RSA_DEV_CSR} \
  -cert ${LOCAL_CA_CERT} \
  -keyfile ${LOCAL_CA_KEY} \
  -outdir /tmp/ \
  -extensions usr_cert \
  -extfile ${RSA_SSL_CONFIG} \
  -batch \
  -out ${RSA_DEV_CERT}


##############
# Create local
# device ECC p256
##############
cd ${LOCAL_CA_DIR}
openssl ecparam -genkey \
    -name prime256v1 \
    -out ${ECC_DEV_KEY} 
openssl req \
    -key ${ECC_DEV_KEY} \
    -new \
    -sha256 \
    -subj "/C=US/ST=CA/O=Example Networks/CN=${ECC_DIDN_ID}" \
    -addext "subjectAltName = DNS:${ECC_DIDN_ID}" \
    -addext "keyUsage = nonRepudiation, digitalSignature, keyEncipherment" \
    -out ${ECC_DEV_CSR}

echo "#################### LOCAL ECC DEV CSR ####################"
openssl req -noout -text -in ${ECC_DEV_CSR}
# Add the SAN config to the end of the openssl conf file.
cp /usr/lib/ssl/openssl.cnf ${ECC_SSL_CONFIG}
echo -e "\n[alternate_names]" >> ${ECC_SSL_CONFIG}
echo -e "DNS.1 = ${ECC_DIDN_ID}\n" >> ${ECC_SSL_CONFIG}
# Accommodate the default behavior of openssl ca.
openssl ca \
  -config ${ECC_SSL_CONFIG} \
  -days 375 \
  -in ${ECC_DEV_CSR} \
  -cert ${LOCAL_CA_CERT} \
  -keyfile ${LOCAL_CA_KEY} \
  -outdir /tmp/ \
  -extensions usr_cert \
  -extfile ${ECC_SSL_CONFIG} \
  -batch \
  -out ${ECC_DEV_CERT}

##############
# Create local
# device ECC p256 IOT REGISTRY
##############
cd ${LOCAL_CA_DIR}
openssl ecparam -genkey \
    -name prime256v1 \
    -out ${IOTREG_DEV_KEY} 
openssl req \
    -key ${IOTREG_DEV_KEY} \
    -new \
    -sha256 \
    -subj "/C=US/ST=CA/O=Example Networks/CN=${IOTREG_CN}" \
    -addext "subjectAltName = DNS:${IOTREG_DIDN_ID}" \
    -addext "keyUsage = nonRepudiation, digitalSignature, keyEncipherment" \
    -out ${IOTREG_DEV_CSR}

echo "#################### LOCAL ECC DEV CSR ####################"
openssl req -noout -text -in ${IOTREG_DEV_CSR}
# Add the SAN config to the end of the openssl conf file.
cp /usr/lib/ssl/openssl.cnf ${ECC_SSL_CONFIG}
echo -e "\n[alternate_names]" >> ${ECC_SSL_CONFIG}
echo -e "DNS.2 = ${IOTREG_DIDN_ID}\n" >> ${ECC_SSL_CONFIG}
# Accommodate the default behavior of openssl ca.
openssl ca \
  -config ${ECC_SSL_CONFIG} \
  -days 375 \
  -in ${IOTREG_DEV_CSR} \
  -cert ${LOCAL_CA_CERT} \
  -keyfile ${LOCAL_CA_KEY} \
  -outdir /tmp/ \
  -extensions usr_cert \
  -extfile ${ECC_SSL_CONFIG} \
  -batch \
  -out ${IOTREG_DEV_CERT}

# cat ${RSA_SSL_CONFIG}
##############
# Copy files
# for export
##############
cp -t ${CRYPTO_EXPORT_PATH} \
    ${LOCAL_CA_KEY} \
    ${LOCAL_CA_CERT} \
    ${RSA_DEV_KEY} \
    ${RSA_DEV_CSR} \
    ${RSA_DEV_CERT} \
    ${ECC_DEV_KEY} \
    ${ECC_DEV_CSR} \
    ${ECC_DEV_CERT} \
    ${IOTREG_DEV_KEY} \
    ${IOTREG_DEV_CSR} \
    ${IOTREG_DEV_CERT}


##############
# Print results
##############
echo "######## RESULTING FILES ##########"
ls -lah ${CRYPTO_EXPORT_PATH}
echo "Crypto builder phase complete!"

echo "####### CERTIFICATE METADATA #######"
echo "## CA certificate:"
openssl x509 -text -noout -in ${LOCAL_CA_CERT}
echo "## Entity certificates:"
echo "#################### LOCAL RSA DEV CERTIFICATE ####################"
openssl x509 -noout -text -in ${RSA_DEV_CERT}

echo "#################### LOCAL RSA DEV CERTIFICATE ####################"
openssl x509 -noout -text -in ${ECC_DEV_CERT}

echo "### END CRYPTO GENERATION PROCESS ###"

tests/scripts/create_testing_trust_map.py
"""Create a testing trust map for pkix_web_verify."""
import argparse
import json

from dane_discovery.dane import DANE

parser = argparse.ArgumentParser()
parser.add_argument("--ssid")
parser.add_argument("--dnsname")
parser.add_argument("--aki")
parser.add_argument("--outfile")
parser.add_argument("--cert_path")

def main():
    args = parser.parse_args()
    with open(args.cert_path) as cert:
        cert_hash = DANE.generate_sha_by_selector(cert.read(), "sha256", 0)
    contents = {args.ssid: {args.dnsname: 
                               {"akis": [args.aki.replace(":", "-").lower()],
                                "cert_hashes": [cert_hash]}}}
    with open(args.outfile, "w") as out:
        out.write(json.dumps(contents))


if __name__ == "__main__":
    main()

tests/integration/test_integration_utility.py
import os
import pprint

from unittest.mock import MagicMock

from radius_pkix_cd.utility import Utility
from dane_discovery.pki import PKI
from dane_discovery.dane import DANE

here_dir = os.path.dirname(os.path.abspath(__file__))
fixtures_dir = os.path.join(here_dir, "../fixtures/")

private_cert_name = "ecc.air-quality-sensor._device.example.net.cert.pem"
private_cert_path = os.path.join("tests/ca2/", private_cert_name)
registry_cert_name = "iotreg.ca.cert.pem"
registry_cert_path = os.path.join("tests/ca2/", registry_cert_name)


class TestIntegrationUtility:
    """Integration tests for the Utility class."""

    def get_file_contents(self, file_path):
        """Return the contents of a file."""
        with open(file_path) as f_p:
            return f_p.read()

    def generate_response(self, msgs):
        """Return an rrset for testing."""
        response = {"dnssec": False, "tcp": True, "tls": False,
                    "responses": msgs}
        return response

    def test_integration_utility_get_authz_config(self):
        infile = os.path.join(fixtures_dir, "trust_source.txt")
        trust_structure = Utility.get_authz_config(infile)
        expected = {"my._device.example.com": ["SSID1"],
                    "your._device.example.com": ["SSID1", "SSID2"]}
        pprint.pprint(trust_structure)
        pprint.pprint(expected)
        assert trust_structure == expected
        return

    def test_integration_check_iot_registry_issuance(self):
        """Test the utility function which identifies IoT Registry issuance."""
        private_cert_pem = self.get_file_contents(private_cert_path)
        registry_cert_pem = self.get_file_contents(registry_cert_path)
        assert Utility.check_iot_registry_issuance(registry_cert_pem)
        assert not Utility.check_iot_registry_issuance(private_cert_pem)

    def test_integration_check_iot_registry_revoked(self):
        """Test the utility function which identifies IoT Registry issuance."""
        tlsa_record_fmt = "{}. 373 IN TLSA {}"
        registry_cert_pem = self.get_file_contents(registry_cert_path)
        non_registry_cert_pem = self.get_file_contents(private_cert_path)
        test_dns_name = PKI.get_cert_meta(registry_cert_pem)["subject"]["commonName"]
        mock_dane = DANE
        
        tlsa_record_good = tlsa_record_fmt.format(test_dns_name, DANE.generate_tlsa_record(3, 0, 1, registry_cert_pem))
        tlsa_record_bad_1 = tlsa_record_fmt.format(test_dns_name, DANE.generate_tlsa_record(3, 0, 1, registry_cert_pem))
        tlsa_record_bad_2 = tlsa_record_fmt.format(test_dns_name, DANE.generate_tlsa_record(3, 1, 1, registry_cert_pem))
        
        response = self.generate_response([tlsa_record_good])
        print(response)
        mock_dane.get_responses = MagicMock(return_value=response)
        result = DANE.get_tlsa_records(test_dns_name)
        # Pass
        assert not Utility.check_iot_registry_revoked(registry_cert_pem, None)
        
        # Fail, hash does not match.
        response = self.generate_response([tlsa_record_bad_1])
        print(response)
        mock_dane.get_responses = MagicMock(return_value=response)
        assert Utility.check_iot_registry_revoked(non_registry_cert_pem, None)
        
        # Fail, bad TLSA record config.
        response = self.generate_response([tlsa_record_bad_2])
        print(response)
        mock_dane.get_responses = MagicMock(return_value=response)
        assert Utility.check_iot_registry_revoked(non_registry_cert_pem, None)

radius_pkix_cd/utility.py
"""Utility functions for radius_pkix_cd."""
import json
import re

from dane_discovery.pki import PKI
from dane_discovery.dane import DANE


class Utility:
    """Various utility functions found here."""

    iot_registry_org_domains = ["iotregistry.ca"]

    @classmethod
    def update_trust_store_file(cls, path, trust_map):
        """Update the trust store file.

        This function will first compare the file
        on disk to what's to be written, and only 
        write the file if it's different than what's
        already on disk.

        Args:
            path (str): Path to trust store file.
            trust_map (dict):This is a dictionary where 
                the first-level key is the realm.
                The value for the first-level key is a 
                dictionary where the key is the DNS
                name and the value is a list of valid
                authorityKeyIDs.
                {"MySSID": 
                    {"my._device.example":
                        {"cert_hashes": 
                            ["cert_hash_1",
                             "cert_hash_2],
                         "akis": 
                            ["AKI1",
                             "AnotherAKI"]
                        }
                    }
                }
        
        Return: True if file was updated, False otherwise.
        """
        try:
            with open(path) as f_on_disk:
                current_map = json.load(f_on_disk)
        except FileNotFoundError:
            current_map = {}
        except json.decoder.JSONDecodeError:
            current_map = {}
        if not current_map == trust_map:
            with open(path, 'w') as f_on_disk:
                json.dump(trust_map, f_on_disk)
                print("Updated trust store at {}".format(path))
                return True
        return False
    
    @classmethod
    def get_authz_config(cls, path):
        """Return dict representing authz configuration.
        
        The format of the file is expected to be pipe- 
        delimited REALM|DNSNAME. If you use pipes in your 
        SSID names... sorry. Also, don't start your SSID 
        with a space. Because that won't work either.

        This function organizes permitted Called-Station-IDs 
        by DNSName.

        If there are incorrectly-formatted lines in the file,
        the problem line number will be printed to stdout and
        the process will continue.
        
        Args:
            path (str): Path to authz file.
            
        Return:
            dict: {"my._device.example": ["MySSID", "OtherSSID"]}
        """
        authz_config = {}
        with open(path) as authz_file:
            line_no = 0
            for line in authz_file.readlines():
                line = line.strip(" ").strip("\n")
                line_no += 1
                if not line:
                    # Catch empty lines
                    continue
                line_parts = line.split("|")
                if len(line_parts) != 2:
                    print("Incorrectly formatted line: {}".format(line_no))
                realm = line_parts[0]
                try:
                    identity_name = line_parts[1]
                    cls.verify_dns_name(identity_name)
                except ValueError as err:
                    print("Bad ID name on line {}: {}".format(line_no, err))
                    continue
                if identity_name not in authz_config:
                    authz_config[identity_name] = [realm]
                else:
                    authz_config[identity_name].append(realm)
        return authz_config

    
    @classmethod
    def verify_dns_name(cls, dns_name):
        """Ensure that ```dns_name`` conforms to RFC 1123 constraints.

        Allowable length: 255 chars.

        Characters: a-z, 1-9, -_.

        Args:
            dns_name (str): DNS name to validate.

        Return:
            None

        Raise:
            ValueError
        """
        dns_name = dns_name.rstrip(".")
        if cls.is_it_an_ip(dns_name):
            errmsg = "'{}' is a bad hostname (is it an IP address?)!".format(dns_name)
            raise ValueError(errmsg)
        split_at_dot = dns_name.split(".")
        pattern = re.compile("^([A-Za-z0-9_-]+)$")
        for x in split_at_dot:
            if not pattern.match(x):
                errmsg = "'{}' is a bad hostname!".format(dns_name)
                raise ValueError("Hostname invalid!")

    @classmethod
    def is_it_an_ip(cls, dns_name):
        """Return True if it's an IP, False otherwise."""
        split_at_dot = dns_name.split(".")
        if len(split_at_dot) != 4:
            return False
        try:
            for x in split_at_dot:
                if not 0 < int(x) < 255:
                    return False
        except ValueError:
            return False
        return True

    @classmethod
    def update_ca_file(cls, file_name, pem_certs):
        """Write all CA certificates to a file."""
        with open(file_name, "wb") as ca_file:
            ca_file.write(b"\n".join(pem_certs))

    @classmethod
    def check_iot_registry_issuance(cls, cert_pem):
        """Return True if the certificate was issued by IoT Registry."""
        cert_meta = PKI.get_cert_meta(cert_pem)
        common_name = cert_meta["subject"]["commonName"]
        for registry_domain in cls.iot_registry_org_domains:
            if cls.dnsname_in_domain(common_name, registry_domain):
                return True
        return False

    @classmethod
    def check_iot_registry_revoked(cls, cert_pem, ns_override=None):
        """Return True if the certificate has been revoked by the IoT registry.
        
        We expect to find exactly one TLSA RR for the identity in the IoT registry.
        
        That TLSA record must be delivered with DNSSEC, and it must be a 
        ``3 0 1`` representation.

        The SHA256 hash must match the presented certificate.
        """
        cert_meta = PKI.get_cert_meta(cert_pem)
        registry_dns_name = cert_meta["subject"]["commonName"]
        registry_entries = DANE.get_tlsa_records(registry_dns_name, nsaddr=ns_override)
        if not registry_entries:
            print("No IoT registry entry for identity!")
            return True
        registry_entry = registry_entries[0]
        reg_txt = str(registry_entry)
        if not (registry_entry["certificate_usage"] == 3 
                and registry_entry["selector"] == 0
                and registry_entry["matching_type"] == 1):
            print("Unexpected TLSA record format from registry: {}".format(reg_txt))
            return True
        expected_sha = DANE.generate_sha_by_selector(cert_pem, "sha256", 0)
        if not expected_sha == registry_entry["certificate_association"]:
            print("Hash mismatch: Registry: {} != Cert: {}".format(registry_entry["certificate_association"], expected_sha))
            return True
        return False
        
    @classmethod
    def domain_str_to_labels(cls, domain_name):
        """Return a list of domain name labels, in reverse-DNS order."""
        labels = domain_name.rstrip(".").split(".")
        labels.reverse()
        return labels

    @classmethod
    def dnsname_in_domain(cls, dns_name, domain_name):
        """Return True if dns_name falls under domain_name, else False.
        
        Forces to lowercase for comparison, since DNS is case-insensitive"""
        dns_name_parts = cls.domain_str_to_labels(dns_name.lower())
        domain_name_parts = cls.domain_str_to_labels(domain_name.lower())
        if len(dns_name_parts) <= len(domain_name_parts):
            return False
        for domain_label in domain_name_parts:
            dns_label = dns_name_parts.pop(0)
            if dns_label != domain_label:
                return False
        return True

radius_pkix_cd/scripts/pkix_cd_verify.py
"""Perform authz based on access configuration and PKIX-CD."""
import argparse
import json
import sys

from dane_discovery.dane import DANE
from dane_discovery.identity import Identity
from dane_discovery.pki import PKI

from radius_pkix_cd.utility import Utility


description=("Authorize supplicants against the access configuration, "
             "using PKIX-CD for identity to trust anchor mapping."
             "\nExit codes:\n"
             "1: Missing trust map.\n"
             "2: Missing certificate file.\n"
             "3: Invalid called-station-id\n"
             "4: Live verify failed.\n"
             "5: IoT Registry check failed.")

parser = argparse.ArgumentParser(description=description)
parser.add_argument("--called", dest="called", required=True, 
                    help="Called-Station-Id.")
parser.add_argument("--calling", dest="calling", required=True, 
                    help="Callling-Station-Id.")
parser.add_argument("--certfile", dest="certfile", required=True, 
                    help="Certificate file presented by supplicant.")
parser.add_argument("--trustmap", dest="trustmap", required=True, 
                    help="Trust map, provided by pkix_cd_manage_trust.")
parser.add_argument("--live-verify", dest="live_verify", required=False, action="store_true", 
                    help="Verify directly against DNS, in addition to cached information.")
parser.add_argument("--require-registry", dest="require_registry", required=False, action="store_true", 
                    help="Set this to require IoT Registry revocation checks for all clients.")
parser.add_argument("--ns_override", dest="ns_override", required=False, help="Override system name server.")
parser.set_defaults(live_verify=False)
parser.set_defaults(require_registry=False)
parser.set_defaults(ns_override=None)



def main():
    """Verify against PKIX-CD and exit according to assertion."""
    args = parser.parse_args()
    # Load the trust map or bail.
    try:
        with open(args.trustmap) as f_on_disk:
            current_map = json.load(f_on_disk)
    except FileNotFoundError:
        print("Trust map not found: {}!".format(args.trustmap))
        sys.exit(1)
    
    # Load the presented certificate or bail.
    try:
        with open(args.certfile) as f_on_disk:
            cert_pem = f_on_disk.read()
    except FileNotFoundError:
        print("Certificate file not found: {}!".format(args.certfile))
        sys.exit(2)
    
    # Make sure that the called station is valid.
    ssid = args.called.split(":", 1)[1]
    if ssid not in current_map:
        print("Invalid called station: {} (no match {})".format(args.called, ssid))
        sys.exit(3)

    # Make sure that the identity is authorized to access the called station.
    if args.calling not in current_map[ssid]:
        print("Identity {} not allowed access to {}".format(args.calling, ssid))

    # Match the certificate's hash against what we have in the trust map.
    cert_hash = DANE.generate_sha_by_selector(cert_pem, "sha256", 0)
    if cert_hash not in current_map[ssid][args.calling]["cert_hashes"]:
        print("Presented certificate does not map to accepted certificate hash!")
        print("{} not in {}".format(cert_hash, current_map[ssid][args.calling]["cert_hashes"]))
        exit(4)

    # Next, we check the cert against the live DNS config!
    if args.live_verify:
        identity = Identity(args.calling, None, args.ns_override)
        success, reason = identity.validate_certificate(cert_pem)
        if not success:
            print("Failed PKIX-CD authentication: {}".format(reason))
            exit(5)

    # Finally, we check for IoT Registry revocation
    issued_by_iotregistry = Utility.check_iot_registry_issuance(cert_pem)
    registry_pass = True
    if args.require_registry and not issued_by_iotregistry:
        print("IoT Registry required, and this identity is not in the IoT Registry.")
        registry_pass = False
    if issued_by_iotregistry and Utility.check_iot_registry_revoked(cert_pem, args.ns_override):
        print("Certificate revoked by IoT Registry!")
        registry_pass = False
    if not registry_pass:
        exit(6)

    # If we've survived to this point, we win!
    exit(0)


if __name__ == "__main__":
    main()

testing_environment/generate_crypto.sh
set -xe
###################################
# This phase creates all the certs
# and keys we'll use in a later
# phase.
###################################

##############
# Set env vars
##############
# Domains
LOCAL_DOMAIN="example.net"

# Base paths
CRYPTO_EXPORT_PATH="${HOME}/export"
CRYPTO_DIR="${HOME}/crypto"
LOCAL_CA_DIR="${CRYPTO_DIR}/local_ca"
DEV_DIR="${CRYPTO_DIR}/device"

# General device info
RSA_DEVICE_SERIAL="rsa"
DEVICE_MODEL="air-quality-sensor"
ECC_DEVICE_SERIAL="ecc"

# Building DIDN-IDs for devices
RSA_DIDN_ID="${RSA_DEVICE_SERIAL}.${DEVICE_MODEL}._device.${LOCAL_DOMAIN}"
ECC_DIDN_ID="${ECC_DEVICE_SERIAL}.${DEVICE_MODEL}._device.${LOCAL_DOMAIN}"

# Specific file paths

## Local CA
LOCAL_CA_KEY="${LOCAL_CA_DIR}/ca.example.net.key.pem"
LOCAL_CA_CERT="${LOCAL_CA_DIR}/ca.example.net.cert.pem"

## RSA Dev
RSA_DEV_KEY="${DEV_DIR}/${RSA_DIDN_ID}.key.pem"
RSA_DEV_CSR="${DEV_DIR}/${RSA_DIDN_ID}.csr.pem"
RSA_DEV_CERT="${DEV_DIR}/${RSA_DIDN_ID}.cert.pem"

## ECC Dev
ECC_DEV_KEY="${DEV_DIR}/${ECC_DIDN_ID}.key.pem"
ECC_DEV_CSR="${DEV_DIR}/${ECC_DIDN_ID}.csr.pem"
ECC_DEV_CERT="${DEV_DIR}/${ECC_DIDN_ID}.cert.pem"

# Signing configurations
RSA_SSL_CONFIG="${DEV_DIR}/${RSA_DIDN_ID}.conf"
ECC_SSL_CONFIG="${DEV_DIR}/${ECC_DIDN_ID}.conf"

mkdir -p ${LOCAL_CA_DIR}/demoCA/
echo "05" > ${LOCAL_CA_DIR}/demoCA/serial 


sudo cp testing_environment/ssl.cnf /usr/lib/ssl/openssl.cnf
# cat /usr/lib/ssl/openssl.cnf



##############
# Create dirs
##############
mkdir -p \
    ${CRYPTO_EXPORT_PATH} \
    ${CRYPTO_DIR} \
    ${LOCAL_CA_DIR} \
    ${DEV_DIR} \
    ${LOCAL_CA_DIR}/demoCA/

##############
# Create local
# CA
##############
cd ${LOCAL_CA_DIR}
openssl genrsa \
    -out ${LOCAL_CA_KEY} \
    4096
openssl req \
    -key ${LOCAL_CA_KEY} \
    -new \
    -x509 \
    -days 7300 \
    -sha256 \
    -subj "/C=US/ST=CA/O=Example Networks/CN=Example Networks CA" \
    -out ${LOCAL_CA_CERT}
openssl x509 -noout -text -in ${LOCAL_CA_CERT}
touch ${LOCAL_CA_DIR}/demoCA/index.txt
touch ${LOCAL_CA_DIR}/demoCA/index.txt.attr

##############
# Create local
# device RSA 2048
##############
cd ${LOCAL_CA_DIR}
openssl genrsa \
    -out ${RSA_DEV_KEY} \
    2048
openssl req \
    -key ${RSA_DEV_KEY} \
    -new \
    -sha256 \
    -subj "/C=US/ST=CA/O=Example Networks/CN=${RSA_DIDN_ID}" \
    -addext "subjectAltName = DNS:${RSA_DIDN_ID}" \
    -addext "keyUsage = nonRepudiation, digitalSignature, keyEncipherment" \
    -out ${RSA_DEV_CSR}

echo "#################### LOCAL RSA DEV CSR ####################"
openssl req -noout -text -in ${RSA_DEV_CSR}
# Add the SAN config to the end of the openssl conf file.
cp /usr/lib/ssl/openssl.cnf ${RSA_SSL_CONFIG}
echo -e "\n[alternate_names]" >> ${RSA_SSL_CONFIG}
echo -e "DNS.1 = ${RSA_DIDN_ID}\n" >> ${RSA_SSL_CONFIG}
# Accommodate the default behavior of openssl ca.
openssl ca \
  -config ${RSA_SSL_CONFIG} \
  -days 375 \
  -in ${RSA_DEV_CSR} \
  -cert ${LOCAL_CA_CERT} \
  -keyfile ${LOCAL_CA_KEY} \
  -outdir /tmp/ \
  -extensions usr_cert \
  -extfile ${RSA_SSL_CONFIG} \
  -batch \
  -out ${RSA_DEV_CERT}


##############
# Create local
# device ECC p256
##############
cd ${LOCAL_CA_DIR}
openssl ecparam -genkey \
    -name prime256v1 \
    -out ${ECC_DEV_KEY} 
openssl req \
    -key ${ECC_DEV_KEY} \
    -new \
    -sha256 \
    -subj "/C=US/ST=CA/O=Example Networks/CN=${ECC_DIDN_ID}" \
    -addext "subjectAltName = DNS:${ECC_DIDN_ID}" \
    -addext "keyUsage = nonRepudiation, digitalSignature, keyEncipherment" \
    -out ${ECC_DEV_CSR}

echo "#################### LOCAL ECC DEV CSR ####################"
openssl req -noout -text -in ${ECC_DEV_CSR}
# Add the SAN config to the end of the openssl conf file.
cp /usr/lib/ssl/openssl.cnf ${ECC_SSL_CONFIG}
echo -e "\n[alternate_names]" >> ${ECC_SSL_CONFIG}
echo -e "DNS.1 = ${ECC_DIDN_ID}\n" >> ${ECC_SSL_CONFIG}
# Accommodate the default behavior of openssl ca.
openssl ca \
  -config ${ECC_SSL_CONFIG} \
  -days 375 \
  -in ${ECC_DEV_CSR} \
  -cert ${LOCAL_CA_CERT} \
  -keyfile ${LOCAL_CA_KEY} \
  -outdir /tmp/ \
  -extensions usr_cert \
  -extfile ${ECC_SSL_CONFIG} \
  -batch \
  -out ${ECC_DEV_CERT}

# cat ${RSA_SSL_CONFIG}
##############
# Copy files
# for export
##############
cp -t ${CRYPTO_EXPORT_PATH} \
    ${LOCAL_CA_KEY} \
    ${LOCAL_CA_CERT} \
    ${RSA_DEV_KEY} \
    ${RSA_DEV_CSR} \
    ${RSA_DEV_CERT} \
    ${ECC_DEV_KEY} \
    ${ECC_DEV_CSR} \
    ${ECC_DEV_CERT}

##############
# Print results
##############
echo "######## RESULTING FILES ##########"
ls -lah ${CRYPTO_EXPORT_PATH}
echo "Crypto builder phase complete!"

echo "####### CERTIFICATE METADATA #######"
echo "## CA certificate:"
openssl x509 -text -noout -in ${LOCAL_CA_CERT}
echo "## Entity certificates:"
echo "#################### LOCAL RSA DEV CERTIFICATE ####################"
openssl x509 -noout -text -in ${RSA_DEV_CERT}

echo "#################### LOCAL RSA DEV CERTIFICATE ####################"
openssl x509 -noout -text -in ${ECC_DEV_CERT}

echo "### END CRYPTO GENERATION PROCESS ###"